subsys/bluetooth/host/id.c

```diff
@@ -1254,6 +1344,151 @@ void bt_id_del(struct bt_keys *keys)
+
+	if (!(keys->flags & BT_KEYS_ID_CONFLICT)) {
+		LOG_DBG("Key has no conflicts for id %u addr %s", id,
+			bt_addr_le_str(&keys->addr));
+		return 0;
+	}
+
+	if (keys->state & BT_KEYS_ID_ADDED) {
+		LOG_DBG("Key is already added to resolving list for id %u addr %s",
+			id, bt_addr_le_str(&keys->addr));
+		return 0;
+	}
+
+	conflict = bt_id_find_conflict(keys, false);
+	if (conflict == NULL) {
+		/* bt_id_find_conflict returns only keys added to the Resolving List (state is
+		 * BT_KEYS_ID_ADDED). If the key has conflict, but no keys were added (for example,
+		 * if the last added key was removed after bt_unpair()), then this function will
+		 * return NULL.
+		 */
+		LOG_WRN("Conflicting key not in the Resolving List.");
+		goto add;
+	}
+
+	__ASSERT_NO_MSG((conflict->flags & BT_KEYS_ID_CONFLICT) != 0);
+
+	LOG_DBG("Found conflicting key with id %u addr %s", conflict->id,
+		bt_addr_le_str(&conflict->addr));
+
+	adv = bt_adv_lookup_by_id(conflict->id);
+	if (adv && atomic_test_bit(adv->flags, BT_ADV_ENABLED)) {
+		LOG_WRN("Cannot remove the conflicting key from the Resolving List while"
+			" advertising");
+		return -EPERM;
+	}
+
+	/* Drop BT_KEYS_ID_PENDING_DEL flag if we were about to delete the keys since we delete it
+	 * here.
+	 */
+	conflict->state &= ~BT_KEYS_ID_PENDING_DEL;
+	bt_id_del(conflict);
+
+add:
+	bt_id_add(keys);
+
+	return 0;
+}
+
+struct bt_id_resolve {
+	uint8_t id;
+	int err;
+};
+
+static void check_and_add_keys_for_id(struct bt_keys *keys, void *data)
+{
+	struct bt_id_resolve *resolve = data;
+
+	if (resolve->err) {
+		/* Skipping other keys because we got error. */
+		return;
+	}
+
+	if (resolve->id != keys->id) {
+		/* We are only interested in keys for the given id */
+		return;
+	}
+
+	resolve->err = conflict_check_and_replace(resolve->id, keys);
+}
+
+/**
+ * @brief Check and update the resolving list for a given identity.
+ *
+ * This function checks if the resolving list contains the keys for the given
+ * identity and peer address. If the keys are not present, it adds them to the
+ * resolving list. If the keys are present, it checks for conflicts with
+ * existing keys in the resolving list. If a conflict is found, it replaces
+ * the conflicting key with the new key.
+ *
+ * If the peer address is NULL, it updates the resolving list for all keys that belong to the given
+ * identity.
+ *
+ * If for any of the keys belonging to the given identity a conflict is found and the advertiser for
+ * that key is enabled, the function returns an error.
+ *
+ * @param id The identity ID to check and update.
+ * @param peer The peer address to check against the resolving list.
+ *
+ * @return 0 on success, or a negative error code on failure.
+ * @return -EPERM if a conflict is found and the advertiser for the conflicting key is enabled.
+ */
+int bt_id_resolving_list_check_and_update(uint8_t id, const bt_addr_le_t *peer)
+{
+	int err;
+
+	if (!IS_ENABLED(CONFIG_BT_ID_AUTO_SWAP_MATCHING_BONDS)) {
+		return 0;
```

Comment:
Should this return error code canceled. Not that it matters, as every call is within option being true. But error code will be good if someone call it by mistake.